### PR TITLE
Add generics and implementations for custom Date and POSIXt escaping

### PR DESCRIPTION
**Pull request by OssiLehtinen**
These modifications address the issue described in https://github.com/tidyverse/dbplyr/issues/384

### CODE REVIEW

**Comment by OssiLehtinen** [comment]
Travis CI was not too happy about my pull request. I'm not familiar with Travis CI so not sure how to proceed.

**Comment by OssiLehtinen** [comment]
I added export of sql_escape_string.

This would enable optionally setting, _e.g._,
```r
sql_escape_string.AthenaConnection <- function(con, x) {
  if(try(as.Date(x), silent=T)  == x) {
    paste0('date ', dbQuoteString(con, x))
  } else {
    dbQuoteString(con, x)
  }
}
```
meaning a call, such as `tbl(con, "calendar") %>% filter(date_col == "2019-01-01")`, would work, as the string gets identified as an isodate and the 'date' prefix gets added.
This is obviously very much subjective, but in pretty much all of our ad-hoc analysis use cases we need to do filtering of data by date, so any streamlining of this would go a long way.

However, I'm uncertain whether automatic detection of isodates should be the default behaviour. This change would enable doing this if wanted.

**Comment by OssiLehtinen** [comment]
A bit of a mess, but I think something like this would actually be needed to have date vectors work out and have only iso-dates cast to dates. Automatically casting something like "1" to a date is not a good thing.

```r
sql_escape_string.AthenaConnection <- function(con, x) {
  all_dates <- all(try(as.Date(x, tryFormats = "%Y-%m-%d"), silent=T) == x)
  if(all_dates & !is.na(all_dates)) {
    paste0('date ', DBI::dbQuoteString(con, x))
  } else {
    DBI::dbQuoteString(con, x)
  }
}
```

In any case, I don't think this should be the default behaviour of dbplyr, but exporting the generic would allow a user to shoot him/herself in the foot, if that is what he/she wishes to do.

**Comment by OssiLehtinen** [comment]
One more note:

if one would define sql_escape_string.AthenaConnection as described above, the escape.Date definition would not be necessary, as currently the dates get cast to strings first and the sql_escape_string function is called on those, which would then detect them as dates. 

However, still I don't think the sql_escape_string should be defined like that as default, so having the Athena-specific escape.Date (and escape.POSIXt) would be a good thing imho.

p.s. If I'm not mixing up things (still learning about s3-classes etc), exporting sql_escape_string is actually not necessary to be able to define the Athena specific version, right?

File: R/escape.R [hunk 1]
```diff
@@ -69,14 +69,13 @@ escape.factor <- function(x, parens = NA, collapse = ", ", con = NULL) {
 
 #' @export
 escape.Date <- function(x, parens = NA, collapse = ", ", con = NULL) {
-  x <- as.character(x)
-  escape.character(x, parens = parens, collapse = collapse, con = con)
+  x = as.character(x)
```
Comment (hadley) [inline R/escape.R:72]: I think you should delete this line; `sql_escape_date()` should be in charge of converting from a date to a character

File: R/escape.R [hunk 2]
```diff
@@ -229,13 +228,37 @@ sql_escape_logical <- function(con, x) {
   UseMethod("sql_escape_logical")
 }
 
+#' @keywords internal
+#' @export
+sql_escape_Date <- function(con, x) {
```
Comment (hadley) [inline R/escape.R:233]: I'd prefer to call these `sql_escape_date()` and `sql_escape_datetime()`.

**Comment by hadley** [comment]
(Also since it's a been a while since you submitted this PR, please let me know if you're not interested, and I'll finish it)

File: R/escape.R [hunk 3]
```diff
@@ -229,13 +228,37 @@ sql_escape_logical <- function(con, x) {
...
 # DBIConnection methods --------------------------------------------------------
 
 #' @export
 sql_escape_string.DBIConnection <- function(con, x) {
   dbQuoteString(con, x)
 }
 
+#' @export
+sql_escape_Date.DBIConnection <- function(con, x) {
+  dbQuoteString(con, x)
```
Comment (krlmlr) [inline R/escape.R:253]: DBI has this, DBI backends can override.

```suggestion
  dbQuoteLiteral(con, x)
```

File: R/escape.R [hunk 4]
```diff
@@ -229,13 +228,37 @@ sql_escape_logical <- function(con, x) {
...
+}
+
+#' @export
+sql_escape_POSIXt.DBIConnection <- function(con, x) {
+  x <- strftime(x, "%Y-%m-%dT%H:%M:%OSZ", tz = "UTC")
+  dbQuoteString(con, x)
```
Comment (krlmlr) [inline R/escape.R:259]: Should this be `%OS6`? DBI needs to be fixed here -- I think passing the time zone might be correct here.

```suggestion
  dbQuoteLiteral(con, x)
```

File: R/escape.R (hunk 1 above)
Comment (OssiLehtinen) [reply R/escape.R]: Moved to coverting bit to sql_escape_date.

File: R/escape.R (hunk 2 above)
Comment (OssiLehtinen) [reply R/escape.R]: Renamed the functions.

File: R/escape.R (hunk 3 above)
Comment (OssiLehtinen) [reply R/escape.R]: I'm sorry, could you explain a bit what would be the benefit of this change? dbQuoteLiteral seems to format a date as something like '2020-04-15 UTC' which does not work at least in Athena (don't have other databases up right now to test against). So if I would use dbQuoteLiteral, I would still need to first cast the date as character.

Surely I'm just missing the point here, so please help me out :)

File: R/escape.R (hunk 4 above)
Comment (OssiLehtinen) [reply R/escape.R:258]: This specific formatting is what is in place in dbplyr at the moment for POSIXt, so I wouldn't want to touch that.

**Comment by OssiLehtinen** [comment]
I made changes to the code according to the recommendations. I hope I got it right. Could you @hadley take a look at the news bullet, as I'm a bit uncertain about the terminology? Thanks!

**Comment by hadley** [comment]
Thanks @OssiLehtinen — I finished it by updating your branch with the latest changes on master, tweaking the news bullet, and making a couple of small documentation and test fixes. Thanks for all your work on this!

**Comment by OssiLehtinen** [comment]
> Thanks @OssiLehtinen — I finished it by updating your branch with the latest changes on master, tweaking the news bullet, and making a couple of small documentation and test fixes. Thanks for all your work on this!

I'm very happy to be able to contribute even a little bit to this great stack of tools!

**Pull request merged**